jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -37,42 +40,33 @@ object XGBoost extends Serializable {
       xgBoostConfMap: Map[String, AnyRef],
       numWorkers: Int, round: Int, obj: ObjectiveTrait, eval: EvalTrait): RDD[Booster] = {
     import DataUtils._
-    val sc = trainingData.sparkContext
-    val tracker = new RabitTracker(numWorkers)
-    if (tracker.start()) {
-      trainingData.repartition(numWorkers).mapPartitions {
-        trainingSamples =>
-          Rabit.init(new java.util.HashMap[String, String]() {
-            put("DMLC_TASK_ID", TaskContext.getPartitionId().toString)
-          })
-          val dMatrix = new DMatrix(new JDMatrix(trainingSamples, null))
-          val booster = SXGBoost.train(xgBoostConfMap, dMatrix, round,
-            watches = new HashMap[String, DMatrix], obj, eval)
-          Rabit.shutdown()
-          Iterator(booster)
-      }.cache()
-    } else {
-      null
-    }
+    trainingData.repartition(numWorkers).mapPartitions {
+      trainingSamples =>
+        Rabit.init(new java.util.HashMap[String, String]() {
+          put("DMLC_TASK_ID", TaskContext.getPartitionId().toString)
+        })
+        val dMatrix = new DMatrix(new JDMatrix(trainingSamples, null))
+        val booster = SXGBoost.train(xgBoostConfMap, dMatrix, round,
+          watches = new mutable.HashMap[String, DMatrix]{put("train", dMatrix)}.toMap, obj, eval)
+        Rabit.shutdown()
+        Iterator(booster)
+    }.cache()
   }
 
-  def train(config: Config, trainingData: RDD[LabeledPoint], obj: ObjectiveTrait = null,
-      eval: EvalTrait = null): Option[XGBoostModel] = {
-    import DataUtils._
-    val numWorkers = config.getInt("numWorkers")
-    val round = config.getInt("round")
+  def train(trainingData: RDD[LabeledPoint], configMap: Map[String, AnyRef], round: Int,
+       obj: ObjectiveTrait = null, eval: EvalTrait = null): Option[XGBoostModel] = {
+    val numWorkers = trainingData.partitions.length
     val sc = trainingData.sparkContext
     val tracker = new RabitTracker(numWorkers)
-    if (tracker.start()) {
-      // TODO: build configuration map from config
-      val xgBoostConfigMap = new HashMap[String, AnyRef]()
-      val boosters = buildDistributedBoosters(trainingData, xgBoostConfigMap, numWorkers, round,
-        obj, eval)
-      // force the job
-      sc.runJob(boosters, (boosters: Iterator[Booster]) => boosters)
-      tracker.waitFor()
-      // TODO: how to choose best model
-      Some(boosters.first())
+    require(tracker.start(), "FAULT: Failed to start tracker")
```

Comment:
Think where we can get numWorkers, also enable default that uses the existing number of partitions in spark